### PR TITLE
Feat/768 manually update cache

### LINKED ISSUE #768: Manually update the cache after a mutation or a subscription
I'm not sure if such thing is possible right now. There's nothing on the docs about it.

I found this issue, on where an `updateCache`  method is proposed, but seems it has not been implemented.

https://github.com/nearform/graphql-hooks/issues/52

### PR DESCRIPTION
**Pull request by mahenrique94**
### What does this PR do?
Resolves #768 issue by creating two new functions (`invalidateQuery` and `setQueryData`) into `GraphqlClient` to update the cache manually after applying some mutation. Here is the difference between both functions:

- `invalidateQuery`: Will delete older cache, re-fetch the new data using the same query, and store in the cache as a new value.
- `setQueryData`: Will override the older cache state with the new one provided by the function return.

### Related issues

- https://github.com/nearform/graphql-hooks/issues/768

### Checklist

- [x] I have checked the [contributing document](../blob/master/CONTRIBUTING.md)
- [x] I have added or updated any relevant documentation
- [x] I have added or updated any relevant tests

### CODE REVIEW

File: packages/graphql-hooks/src/GraphQLClient.ts [hunk 1]
```diff
@@ -200,6 +201,12 @@ class GraphQLClient {
     }
   }
 
+  removeCache(cacheKey) {
+    if (this.cache) {
```
Comment (ilteoood) [inline packages/graphql-hooks/src/GraphQLClient.ts:205]: we can remove the if doing just:
```
this.cache?.delete(cacheKey)
```

File: packages/graphql-hooks/src/GraphQLClient.ts [hunk 2]
```diff
@@ -407,6 +414,31 @@ class GraphQLClient {
...
     }
   }
+
+  invalidateQuery(query: string): void {
+    const cacheKey = this.getCacheKey({ query })
+    if (this.cache && cacheKey) {
+      this.removeCache(cacheKey)
+      this.request({ query }).then(result =>
+        this.mutationsEmitter.emit(Events.DATA_INVALIDATED, result)
+      )
```
Comment (ilteoood) [inline packages/graphql-hooks/src/GraphQLClient.ts:424]: Should we catch possible errors?

File: packages/graphql-hooks/src/GraphQLClient.ts [hunk 3]
```diff
@@ -407,6 +414,31 @@ class GraphQLClient {
...
+  }
+
+  setQueryData(query: string, updater: (oldState?: any) => any): void {
+    const cacheKey = this.getCacheKey({ query })
+    if (this.cache && cacheKey) {
+      const oldState: any = this.cache.get(cacheKey)
+      const newState = {
+        ...oldState,
+        data: {
+          ...updater(oldState.data || null)
```
Comment (ilteoood) [inline packages/graphql-hooks/src/GraphQLClient.ts:435]: as we are already creating a new object here, is this spread necessary?
can we do just
```
data: updater(...)
```
What do you think?

File: packages/graphql-hooks/src/useClientRequest.ts [hunk 4]
```diff
@@ -246,6 +247,17 @@ function useClientRequest<
           })
         }
 
+        if (
+          revisedOpts.onSuccess &&
```
Comment (ilteoood) [inline packages/graphql-hooks/src/useClientRequest.ts:251]: undefined is not an instance of Function, should we remove this check?

**Review by ilteoood — CHANGES_REQUESTED** [review]

File: packages/graphql-hooks/src/useClientRequest.ts (hunk 4 above)
Comment (mahenrique94) [reply packages/graphql-hooks/src/useClientRequest.ts]: made in https://github.com/nearform/graphql-hooks/pull/928/commits/4ea3bd85457943db1b1032920ce4d5d03901d171

File: packages/graphql-hooks/src/GraphQLClient.ts (hunk 3 above)
Comment (mahenrique94) [reply packages/graphql-hooks/src/GraphQLClient.ts]: made in https://github.com/nearform/graphql-hooks/pull/928/commits/4ea3bd85457943db1b1032920ce4d5d03901d171

File: packages/graphql-hooks/src/GraphQLClient.ts (hunk 1 above)
Comment (mahenrique94) [reply packages/graphql-hooks/src/GraphQLClient.ts]: made in https://github.com/nearform/graphql-hooks/pull/928/commits/4ea3bd85457943db1b1032920ce4d5d03901d171

File: examples/persisted-queries/server/mercurius.ts [hunk 5]
```diff
@@ -34,7 +34,11 @@ app.register(mercurius, {
 
 async function start() {
   try {
-    return app.listen({ port: process.env.SERVER_PORT ? parseInt(process.env.SERVER_PORT, 10) : 8000 })
+    return app.listen({
```
[the remark this answers is not in the document]
Comment (mahenrique94) [reply examples/persisted-queries/server/mercurius.ts:37]: made in https://github.com/nearform/graphql-hooks/pull/928/commits/4ea3bd85457943db1b1032920ce4d5d03901d171

**Review by radomird — APPROVED** [review]
LGTM

**Comment by mahenrique94** [comment]
@radomird @ilteoood the PR as ready to review, could you please take a look again?

**Comment by mahenrique94** [comment]
for some reason after rebasing some tests start failing, your changes could collaborate for this to happen @ilteoood?

**Comment by ilteoood** [comment]
> for some reason after rebasing some tests start failing, your changes could collaborate for this to happen @ilteoood?

Of course, starting with React 18 `@testing-library/react-hooks` has been deprecated, you have to use `@testing-library/react` which exposes the same functions.

Start with that, then we'll the other failing tests.

**Comment by mahenrique94** [comment]
thank you @ilteoood, could you review it, please?

**Review by ilteoood — APPROVED** [review]
LGTM

**Pull request merged**